docs/how-tos/build-a-plugin.md

```diff
@@ -0,0 +1,385 @@
+
+Example:
+
+```json
+{
+  "/entry/instrument/source/type": "None"
+}
+```
+
+**Units**: If there is a field defined in the NXDL, the converter expects a filled in /data/@units entry in the template dictionary corresponding to the right /data field unless it is specified as NX_UNITLESS in the NXDL. Otherwise, you will get an exception.
+
+```json
+{
+  "/ENTRY[my_entry]/INSTRUMENT[my_instrument]/SOURCE[my_source]/data": "None",
+  "/ENTRY[my_entry]/INSTRUMENT[my_instrument]/SOURCE[my_source]/data/@units": "Should be set to a string value"
+}
+```
+
+In case the NXDL does not define a `name` for the group the requested data belongs to, the template dictionary will list it as `/NAME_IN_NXDL[name_in_output_nexus]`
+You can choose any name you prefer instead of the suggested name. This allows the reader function to repeat groups defined in the NXDL to be outputted to the NeXus file.
+
+```json
+{
+  "/ENTRY[my_entry]/INSTRUMENT[my_instrument]/SOURCE[my_source]/type": "None"
+}
+```
+
+For attributes defined in the NXDL, the reader template dictionary will have the assosciated key with a "@" prefix to the attributes name at the end of the path:
+
+```json
+{
+  "/entry/instrument/source/@attribute": "None"
+}
+```
+
+You can also define links by setting the value to sub dictionary object with key `link`:
+
+```python
+template["/entry/instrument/source"] = {"link": "/path/to/source/data"}
+```
+
+For a given NXDL schema, you can generate an empty template with the command
+```console
+user@box:~$ dataconverter generate-template` --nxdl NXmynxdl
+```
+
+## How to use the built-in MultiFormatReader
+While building on the ```BaseReader``` allows for the most flexibility, in most cases it is desirable to implement a reader that can read in multiple file formats and then populate the template based on the read data. For this purpose, `pynxtools` has the [**`MultiFormatReader`**](https://github.com/FAIRmat-NFDI/pynxtools/blob/master/src/pynxtools/dataconverter/readers/multi/reader.py), which can be readily extended for your own data.
+
+In case you want to make use of the `MultiFormatReader`, the following basic structure must be implemented:
+```python
+"""MyDataReader implementation for the DataConverter to convert mydata to NeXus."""
+from typing import Tuple, Any
+
+from pynxtools.dataconverter.readers.base.reader import ParseJsonCallbacks, MultiFormatReader
+
+class MyDataReader(MultiFormatReader):
+    """MyDataReader implementation for the DataConverter to convert mydata to NeXus."""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.extensions = {
+            ".yml": self.handle_eln_file,
+            ".yaml": self.handle_eln_file,
+            ".json": self.set_config_file,
+            # Here, you can add functions for handling any other file extension(s)
+        }
+# This has to be set to allow the convert script to use this reader. Set it to "MyDataReader".
+READER = MyDataReader
+```
```

Comment:
I advise against explaining this with the original reader code. It's rather complicated in comparison on how the reader is constructed (i.e., the complicated code is in the multiformatreader such that they can be implemented easier).

I would also add the functions you put into the dict here to explain it with a simple example.

Reply:
I would maybe give an example and then construct a reader for it around it. Expanding the reader step by step (first adding just an eln handling, then a `get_data` function etc etc.)

Reply:
I decided to split these docs into a learning section (using the existing parts) and a how-to along the step by step idea described here.